js/ui/animation.js

```diff
@@ -99,17 +116,31 @@ var Animation = new Lang.Class({
         this.stop();
     },
 
-    setFrameTimeouts: function(timeoutList) {
-        if (timeoutList.length == 0)
-            throw new Error('Cannot set an empty list as the frame-timeouts for animation!');
+    _parseFrame: function(frameInfo) {
+        if (typeof frameInfo === 'string') {
+            let [frame, delay] = frameInfo.split(' ');
+            frame = parseInt(frame);
 
+            if (delay.indexOf('-') > -1)
+                delay = parseInt(delay);
+
+            return [frame, delay];
+        }
+        return [frameInfo, this._speed];
+    },
+
+    setFrames: function(framesList) {
         let wasPlaying = this._isPlaying;
         this.stop();
 
-        this._frameTimeouts = timeoutList;
+        this._frames = [];
+        for (let frameInfo of framesList) {
+            let [frame, delay] = this._parseFrame(frameInfo);
```

Comment:
See comment about having this outside of the `Animation` class.

Reply:
+1